### PR TITLE
Add append function for appending only part of a trajectory

### PR DESCRIPTION
**Pull request by martiniil**
### Description

As a solution to #1212, we propose adding an append function to robot_trajectory::RobotTrajectory, which takes a start- and end-index for appending only a part of the trajectory. By this, we avoid re-assembling or removing points from an existing trajectory.

A cherry-pick to kinetic would be desirable.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Document API changes relevant to the user in the moveit/MIGRATION.md notes(http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [x] Decide if this should be cherry-picked to other current ROS branches

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by simonschmeisser** [comment]
please remove work items (checkboxes) that you don't intent to ever do (ie screenshots etc)

Otherwise this is a really nice PR and I'm looking forward to using it!

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp [hunk 1]
```diff
@@ -93,6 +93,21 @@ void RobotTrajectory::append(const RobotTrajectory& source, double dt)
     duration_from_previous_[index] += dt;
 }
 
+void RobotTrajectory::append(const RobotTrajectory& source, double dt, size_t start_index, size_t end_index)
```
Comment (simonschmeisser) [inline moveit_core/robot_trajectory/src/robot_trajectory.cpp]: Can "end_index" be guessed for convenience? `end_index = 0` leads to `end_index = source.waypoints_.size()`

File: MIGRATION.md [hunk 2]
```diff
@@ -8,6 +8,7 @@ API changes in MoveIt! releases
 - The move_group capability ``ExecuteTrajectoryServiceCapability`` has been removed in favor of the improved ``ExecuteTrajectoryActionCapability`` capability. Since Indigo, both capabilities were supported. If you still load default capabilities in your ``config/launch/move_group.launch``, you can just remove them from the capabilities parameter. The correct default capabilities will be loaded automatically.
 - Deprecated method ``CurrentStateMonitor::waitForCurrentState(double wait_time)`` was finally removed.
 - Renamed ``RobotState::getCollisionBodyTransforms`` to ``getCollisionBodyTransform`` as it returns a single transform only.
+- Added ``RobotTrajectory::append(const RobotTrajectory& source, double dt, size_t start_index, size_t end_index)``.
```
Comment (rhaschke) [inline MIGRATION.md]: I think this should not be part of migration notes, but rather will become part of the changelog.

**Comment by rhaschke** [comment]
Please, _rebase_ to resolve conflicts.

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp (hunk 1 above)
Comment (martiniil) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: I am not sure, if I get your point. Are you talking about default arguments? start_index=0 would make sense, but end_index=0 should lead to an empty append trajectory in my opinion. I think that is more intuitive.

File: MIGRATION.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (martiniil) [reply MIGRATION.md]: I see...

Comment (martiniil) [reply MIGRATION.md]: Ok, thanks.

File: moveit_core/robot_trajectory/src/robot_trajectory.cpp (hunk 1 above)
Comment (rhaschke) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: What about `end_index = std::numeric_limits<std::size_t>::max()`?

Comment (rhaschke) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: As soon as we have default args, for both `start_index` and `end_index`, you can drop the old `append()` method completely.

Comment (martiniil) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: Sounds good.

Comment (v4hn) [reply moveit_core/robot_trajectory/src/robot_trajectory.cpp]: > What about `end_index = std::numeric_limits<std::size_t>::max()`?

0 would have been just fine here as a special case.
It does not make sense to ask to copy no element at all.
But the `max` solution is fine to.

**Comment by v4hn** [comment]
https://github.com/ros-planning/moveit/pull/1162 provides a more general approach and will hopefully allow iterators here later, not just size_t indices.

The approach taken here is valid though and I will merge it.

**Comment by v4hn** [comment]
@martiniil Please feel invited to support the WIP pull-request then :-)
It looks like the original author does not have enough time to work on it after WMD.

**Pull request merged**